### PR TITLE
Created custom version of Ruby’s enumerable methods

### PR DESCRIPTION
**Pull request by Mhdez221993**
In this project we re-create the following enumerables :
1. #my_each
2. #my_each_with_index
3. #my_select
4. #my_all?
5. #my_any?
6. #my_none?
7. #my_count
8. #my_map
9. #my_inject
10. #multiply_els

we use git-flow and make use of the yield method
we follow the project specifications

### CODE REVIEW

File: enumerable.rb [hunk 1]
```diff
@@ -1,87 +1,163 @@
...
+def multiply_els(args)
+  args.my_inject(:*)
+end
 
-# p [ nil, true, 99].my_all?  
-# p [nil, true, 99].all?  
+p [1, 3, 4, 5].my_inject(2, &:+)
 
-# p [].my_all?
-# p [].all?
```
Comment (Berabjesus) [inline enumerable.rb]: - [x] Remove debugging instructions and comments

File: enumerable.rb [hunk 2]
```diff
@@ -1,87 +1,163 @@
...
-        self.class == Array ? yield(i) : yield(i[0], i[1])
+    return to_enum(:my_each) unless block_given?
+
+    i = 0
+    while i < to_a.length
+      instance_of?(Array) ? yield(to_a[i]) : yield(to_a[i][0], to_a[i][1])
+      i += 1
     end
+    self
   end
```
Comment (Berabjesus) [inline enumerable.rb]: - [x]  when a block is given when self is a range behaves like #each calls the given block once for each element in self

```
  range = Range.new(1,5)
  my_each_output = ''
  block =  proc { |num| my_each_output += num.to_s }
  range.each(&block)
  p my_each_output # this is the wrong output. Test with the original each 
```

File: enumerable.rb [hunk 3]
```diff
@@ -1,87 +1,163 @@
...
-      self.class == Array ? yield(i, self.index(i)) : yield([i[0], i[1]], self.keys.index(i[0]))
+    return to_enum(:my_each_with_index) unless block_given?
+
+    i = 0
+    while i < to_a.length
+      instance_of?(Array) ? yield(to_a[i], i) : yield([to_a[i][0], to_a[i][1]], i)
+      i += 1
     end
+    self
   end
```
Comment (Berabjesus) [inline enumerable.rb]: - [x] when a block is given when self is a range behaves like #each_with_index calls the given block once for each element in self.
```
  range = Range.new(1,5)
  my_each_output = ''
  block =  proc { |num| my_each_output += num.to_s }
  range.my_each_with_index(&block) # check with range.each_with_index(&block) 
  p my_each_output  # wrong output 
```

File: enumerable.rb [hunk 4]
```diff
@@ -1,87 +1,163 @@
...
-        end
+    my_each do |k, v|
+      if instance_of?(Array)
+        new_arr << k if yield(k) == true
+      elsif yield(k, v) == true
+        new_hash.store(k, v)
+      end
     end
     new_arr.empty? ? new_hash : new_arr
   end
```
Comment (Berabjesus) [inline enumerable.rb]: - [ ]returns an array containing all elements of enum for which the given block returns a true value::range
```
  range = Range.new(1,10)
  block =  proc { |num| num < (0 + 9) / 2 }
  p range.my_select(&block)  # gotten {1=>0, 0=>1}

  but
  
  range = Range.new(1,10)
  block =  proc { |num| num < (0 + 9) / 2 }
  p range.select(&block)  # is [1, 2, 3]
```

File: enumerable.rb [hunk 5]
```diff
@@ -1,87 +1,163 @@
...
-          if ( args[0].class == Regexp) &&(args[0].match(e.to_s)) == nil
-            return false
-          end
-          if ( args[0].class == Class) && (e.is_a? args[0]) == false 
-            return false
-          end
+    unless block_given?
+      if index(nil)
+        return !index(nil) ? true : false
+      end
```
Comment (Berabjesus) [inline enumerable.rb:54]: - [x] when no block or argument is given returns false when one of the collection members is false or nil
```
  p [1, false, 'hi', []].all?  # false
  p [1, false, 'hi', []].my_all? # true
```

- [x] when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
```
  p [3,3,3,3].all?(3) # true
  p [3,3,3,3].my_all?(3) # False
```

File: enumerable.rb [hunk 6]
```diff
@@ -1,87 +1,163 @@
...
-# p [nil, nil, nil].all?
-
-# p ["some", "somithimes", "something"].my_all?(/s/)
-# p ["some", "somithimes", "something"].all?(/s/)
+        my_each { |v| return true if v == args[0] } == false
+      end
+    else
+      my_each { |v| return true if yield(v) } == false
+    end
+  end
```
Comment (Berabjesus) [inline enumerable.rb]: - [x] returns false if the block ever returns a value false or nil::range
```
  block = proc { |num| num > 9 }
  range = Range.new(5, 50)
  p range.my_any?(&block) # false
  p range.any?(&block) # true
```

- [x]  when no block or argument is given returns false if at least one of the collection is not true
```
  p [nil, false, nil, false].any? # false
  p [nil, false, nil, false].my_any? # true
```

**Review by Berabjesus — CHANGES_REQUESTED** [review]
## Status: CHANGES REQUIRED :warning: :construction: 
Hi @Mhdez221993  and @ShinobiWarior  a very awesome start :+1: Most of the required functionalities work 👏 

#### However, since this is a very important section in ruby, there is a detailed test for it and you do have a few changes to carry out before you proceed so please take a look at the things you have to work on and their respective review comments below.

- [x] Remove debugging instructions

### my_each
- [x] when a block is given when self is a range behaves like #each calls the given block once for each element in self
- [x]  when a block is given when self is a hash behaves like #each calls the given block once for each element in self (check comment below)

### my_each_with_index
- [x]  when a block is given when self is a range returns the range itself after calling the given block once for each element in self
- [x]  when a block is given when self is a range behaves like #each_with_index calls the given block once for each element in self

### my_select
- [x]  returns an array containing all elements of enum for which the given block returns a true value::range

### my_all?
- [x]  when no block or argument is given returns false when one of the collection members are false or nil
- [x]  when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern

### my_any?
- [x]  returns false if the block ever returns a value false or nil::range
- [x] when no block or argument is given returns false if at least one of the collection is not true. 

### my_none? 
- [x] when no block or argument is given returns true only if none of the collection members is true

### my_inject
- [x]  raises a "LocalJumpError" when no block or argument is given. Your method It is giving TypeError instead.

##**If you have any questions, let me know by leaving a comment here and mentioning me. I will be happy to help** 
**HAPPY CODING :computer:[Bereket Beshane](https://bereketababu.me)** | [LinkedIn](https://linkedin.com/in/bereketbeshane) | [GitHub](https://github.com/Berabjesus)

File: enumerable.rb (hunk 2 above)
Comment (Mhdez221993) [reply enumerable.rb]: Thank you for the test code, it helps a lot.

File: enumerable.rb (hunk 1 above)
Comment (Mhdez221993) [reply enumerable.rb]: Thank you for the test code, it helps a lot.

**Review by DeliceLydia — CHANGES_REQUESTED** [review]
# Status: Change Requested

Amazing job @Mhdez221993 and @ShinobiWarior I know this project is not an easy one but you almost there you implemented the change requested by the previous TSE but there is what you need to work on before moving on

## To improve image

- Enumerable #my_map executes only the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum.
- Enumerable #my_inject searches for the longest word in an array of strings
- Enumerable #my_inject does not mutate the original array
- Enumerable #my_inject when a block is given without an initial value combines all elements of enum by applying a binary operation, specified by a block

## KIndly work on that and submit another CRR

Regards,

Lydia Ingabire
Slack: Delice Lydia
[Linkedin](https://www.linkedin.com/in/delice-lydia/)

**Comment by DeliceLydia** [comment]
my_inject

- My_inject find the longest word
longest = %w{ cat sheep bear }.inject do |memo, word|
   memo.length > word.length ? memo : word
end

- enumerable #my_inject does not mutate the original array
- enumerable #multiply_els accepts an array as an argument and multiplies all the elements of the array together using #my_inject
`def multiply_els(array)
  array.my_inject { |product, i| product * i }
end`

**Comment by DeliceLydia** [comment]
enumerable #my_inject when a block is given without an initial value combines all elements of enum by applying a binary operation, specified by a block::range

`(5..10).my_inject { |sum, n| sum + n } `

**Comment by DeliceLydia** [comment]
enumerable #my_map executes only the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum

`my_proc = proc { |i| i * 2 }`
`(1..4).my_map(my_proc) { |i| i*i } `

File: enumerable.rb [hunk 7]
```diff
@@ -1,87 +1,146 @@
...
+
+    i = 0
+    while i < to_a.length
+      instance_of?(Hash) ? yield([to_a[i][0], to_a[i][1]], i) : yield(to_a[i], i)
+      i += 1
     end
+    self
   end
 
   def my_select
```
Comment (Berabjesus) [inline enumerable.rb:30]: - [ ] my_select returns an **array** containing all elements of enum for which the given block returns a true value::range. 

```
  range = Range.new(5,50)
  block = proc { |num| num < (0 + 9) / 2 }
  p range.my_select(&block) // returns { }
  p range.select(&block) // returns [ ]
```

**Review by Berabjesus — APPROVED** [review]
# **Status: PROJECT APPROVED** 
#### Excellent job on this project @Mhdez221993  and @ShinobiWarior . :fireworks: Keep it up.  :champagne: :champagne:
 
**All the required tasks have been fulfilled in this project. I will go ahead and mark this as complete. 👍** 

### **Before you merge to the default branch please have a look at the following OPTIONAL comments below.**

- [x] [OPTIONAL] Remove the `app_screenshot.png` file from your repo;
- [ ] [OPTIONAL] #my_select returns an **array** containing all elements of enum for which the given block returns a true value::range. (check comment below)

I wish you all the very best in your next tasks:1st_place_medal:
#### **If you have any questions, let me know by leaving a comment here and mentioning me. I will be happy to help** 
**HAPPY CODING :computer:[Bereket Beshane](https://bereketababu.me)** | [LinkedIn](https://linkedin.com/in/bereketbeshane) | [GitHub](https://github.com/Berabjesus)

**Pull request merged**